third_party/move/move-model/bytecode/src/dataflow_domains.rs

```diff
@@ -310,3 +309,105 @@ impl<K: Ord + Clone, V: AbstractDomain + Clone + PartialEq> MapDomain<K, V> {
         self.extend(new_values);
     }
 }
+
+impl<A: AbstractDomain, B: AbstractDomain> AbstractDomain for (A, B) {
+    fn join(&mut self, other: &Self) -> JoinResult {
+        let mut join_res = JoinResult::Unchanged;
+        join_res = JoinResult::combine(join_res, self.0.join(&other.0));
+        join_res = JoinResult::combine(join_res, self.1.join(&other.1));
+        join_res
+    }
+}
+
+impl<T: AbstractDomain> AbstractDomain for Vec<T> {
+    fn join(&mut self, other: &Self) -> JoinResult {
+        self.iter_mut()
+            .zip(other.iter())
+            .fold(JoinResult::Unchanged, |acc, (x, y)| {
+                JoinResult::combine(acc, x.join(y))
+            })
+    }
+}
+
+/// Add a formal top and bottom element to type `T`
+#[derive(PartialEq, Eq, Clone)]
+pub enum Plus2<T> {
```

Comment:
Can you please land the other PR first, then I can review here. 

As I also said on the other PR, I'm not sure whether we need this type. This is the first time I see the need for it. Sure its a well-known fact you can create a pseudo-lattice with this approach, but I don't see is commonly needed.

Reply:
> Can you please land the other PR first, then I can review here.
> 
> As I also said on the other PR, I'm not sure whether we need this type. This is the first time I see the need for it. Sure its a well-known fact you can create a pseudo-lattice with this approach, but I don't see is commonly needed.

@wrwg I thought about adding this because I saw in Vineeth's code, but I didn't add it at the time
```
impl AbstractDomain for Initialized {
    /// Implements `join` for the initialized state lattice:
    ///
    ///             +-------+
    ///             | Maybe |
    ///             +-------+
    ///               /   \
    ///              /     \
    ///      +-----+/       \+----+
    ///      | Yes |         | No |
    ///      +-----+\       /+----+
    ///              \     /
    ///               \   /
    ///            +--------+
    ///            | bottom |
    ///            +--------+
    /// Note that bottom is not explicitly represented in the code.
    ```
And I added this type only after I found myself needing this again in my code in this PR.

Reply:
I asked Vineeth to extract this part of the PR out, and it will be reviewed later, and did not land yet.

I need to look closer at your use cases, but in general if you need a full lattice instead of just a semi-lattice for fixed point building sounds like a red flag to me. Only if your program analysis domain would be shared between backward/forward this is needed. And indeed, critical edges removed is assumed.

Reply:
done. (removed)